synapse/util/caches/lrucache.py

```diff
@@ -82,19 +93,243 @@ def enumerate_leaves(node, depth):
                 yield m
 
 
+class _CacheEntry(Protocol):
+    """A protocol for cache entries used by `_ListNode` that allows it to cause
+    the cache entry to be dropped.
+    """
+
+    def drop_from_cache(self) -> None:
+        """Request the entry be dropped from the cache.
+
+        Note: This should call `.remove_from_list()` on all `_ListNodes`
+        referencing it.
+        """
+        ...
+
+
+P = TypeVar("P", bound=_CacheEntry)
+
+
+class _ListNode(Generic[P]):
+    """A node in a doubly linked list, with an (optional) weak reference to a
+    cache entry.
+
+    The weak reference should only be `None` for the root node.
+    """
+
+    # We don't use attrs here as in py3.6 you can't have `attr.s(slots=True)`
+    # and inherit from `Generic` for some reason
+    __slots__ = [
+        "cache_entry",
+        "prev_node",
+        "next_node",
+    ]
+
+    def __init__(self, cache_entry: Optional[P] = None) -> None:
+        self.cache_entry = cache_entry
+        self.prev_node: Optional[_ListNode[P]] = self
+        self.next_node: Optional[_ListNode[P]] = self
+
+    @staticmethod
+    def insert_after(
+        cache_entry: P, root: "_ListNode[P]", clock: Clock
+    ) -> "_ListNode[P]":
```

Comment:
I find the parameter ordering here a bit unintuitive. I'd expect `root` and `cache_entry` to be the other way around.